File: USRadioApp/app/src/main/java/com/usradio/app/adapter/RadioStationAdapter.java
package com.usradio.app.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.usradio.app.R;
import com.usradio.app.model.RadioStation;
import java.util.ArrayList;
import java.util.List;

public class RadioStationAdapter extends RecyclerView.Adapter<RadioStationAdapter.ViewHolder> {
    
    private List<RadioStation> stations = new ArrayList<>();
    private OnStationClickListener listener;
    
    public interface OnStationClickListener {
        void onStationClick(RadioStation station);
        void onFavoriteClick(RadioStation station);
    }
    
    public RadioStationAdapter(OnStationClickListener listener) {
        this.listener = listener;
    }
    
    public void setStations(List<RadioStation> stations) {
        this.stations = stations != null ? stations : new ArrayList<>();
        notifyDataSetChanged();
    }
    
    public void updateStation(RadioStation station) {
        for (int i = 0; i < stations.size(); i++) {
            if (stations.get(i).getStationUuid().equals(station.getStationUuid())) {
                stations.set(i, station);
                notifyItemChanged(i);
                break;
            }
        }
    }
    
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.item_radio_station, parent, false);
        return new ViewHolder(view);
    }
    
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        RadioStation station = stations.get(position);
        holder.bind(station);
    }
    
    @Override
    public int getItemCount() {
        return stations.size();
    }
    
    class ViewHolder extends RecyclerView.ViewHolder {
        ImageView imgLogo;
        TextView txtName;
        TextView txtState;
        TextView txtTags;
        TextView txtBitrate;
        ImageButton btnFavorite;
        
        ViewHolder(View itemView) {
            super(itemView);
            imgLogo = itemView.findViewById(R.id.imgStationLogo);
            txtName = itemView.findViewById(R.id.txtStationName);
            txtState = itemView.findViewById(R.id.txtStationState);
            txtTags = itemView.findViewById(R.id.txtStationTags);
            txtBitrate = itemView.findViewById(R.id.txtStationBitrate);
            btnFavorite = itemView.findViewById(R.id.btnFavorite);
        }
        
        void bind(RadioStation station) {
            txtName.setText(station.getName() != null ? station.getName() : "Unknown Station");
            
            String stateInfo = station.getState() != null && !station.getState().isEmpty() 
                ? station.getState() 
                : "USA";
            txtState.setText(stateInfo);
            
            String tags = station.getTags() != null && !station.getTags().isEmpty() 
                ? station.getTags() 
                : "No tags";
            txtTags.setText(tags);
            
            String bitrateInfo = station.getBitrate() > 0 
                ? station.getBitrate() + " kbps" 
                : "Unknown bitrate";
            if (station.getCodec() != null && !station.getCodec().isEmpty()) {
                bitrateInfo += " • " + station.getCodec().toUpperCase();
            }
            txtBitrate.setText(bitrateInfo);
            
            // Load favicon/logo
            if (station.getFavicon() != null && !station.getFavicon().isEmpty()) {
                Glide.with(itemView.getContext())
                    .load(station.getFavicon())
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .placeholder(android.R.drawable.ic_btn_speak_now)
                    .error(android.R.drawable.ic_btn_speak_now)
                    .into(imgLogo);
            } else {
                imgLogo.setImageResource(android.R.drawable.ic_btn_speak_now);
            }
            
            // Set favorite icon
            if (station.isFavorite()) {
                btnFavorite.setImageResource(android.R.drawable.btn_star_big_on);
            } else {
                btnFavorite.setImageResource(android.R.drawable.btn_star_big_off);
            }
            
            // Click listeners
            itemView.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onStationClick(station);
                }
            });
            
            btnFavorite.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onFavoriteClick(station);
                }
            });
        }
    }
}

File: USRadioApp/app/src/main/java/com/usradio/app/adapter/RadioStationAdapter.kt
package com.usradio.app.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.usradio.app.R
import com.usradio.app.model.RadioStation

class RadioStationAdapter(
    private val onPlayClick: (RadioStation) -> Unit,
    private val onFavoriteClick: (RadioStation) -> Unit
) : ListAdapter<RadioStation, RadioStationAdapter.ViewHolder>(DiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_radio_station, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val stationLogo: ImageView = itemView.findViewById(R.id.stationLogo)
        private val stationName: TextView = itemView.findViewById(R.id.stationName)
        private val stationCountry: TextView = itemView.findViewById(R.id.stationCountry)
        private val stationTags: TextView = itemView.findViewById(R.id.stationTags)
        private val favoriteButton: ImageButton = itemView.findViewById(R.id.favoriteButton)
        private val playButton: ImageButton = itemView.findViewById(R.id.playButton)

        fun bind(station: RadioStation) {
            stationName.text = station.getDisplayName()
            stationCountry.text = station.getDisplayCountry()
            stationTags.text = station.getDisplayTags()

            // Load station logo
            val faviconUrl = station.getFaviconUrl()
            if (faviconUrl != null) {
                Glide.with(itemView.context)
                    .load(faviconUrl)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .placeholder(R.drawable.ic_radio_default)
                    .error(R.drawable.ic_radio_default)
                    .into(stationLogo)
            } else {
                stationLogo.setImageResource(R.drawable.ic_radio_default)
            }

            // Set favorite button state
            favoriteButton.setImageResource(
                if (station.isFavorite) R.drawable.ic_favorite else R.drawable.ic_favorite_border
            )

            // Set click listeners
            playButton.setOnClickListener { onPlayClick(station) }
            favoriteButton.setOnClickListener { onFavoriteClick(station) }
            
            itemView.setOnClickListener { onPlayClick(station) }
        }
    }

    private class DiffCallback : DiffUtil.ItemCallback<RadioStation>() {
        override fun areItemsTheSame(oldItem: RadioStation, newItem: RadioStation): Boolean {
            return oldItem.stationUuid == newItem.stationUuid
        }

        override fun areContentsTheSame(oldItem: RadioStation, newItem: RadioStation): Boolean {
            return oldItem == newItem
        }
    }
}